contracts/protocol/bases/PriceDiscoveryBase.sol

```diff
@@ -62,124 +72,250 @@ contract PriceDiscoveryBase is ProtocolBase {
+        clearStorage(_tokenId);
     }
 
     /**
-     * @notice Fulfils a bid order on external contract.
+     * @notice Fulfils a bid order on external contract when caller is not voucher owner.
      *
      * Reverts if:
-     *  - Voucher owner did not approve protocol to transfer the voucher
-     *  - Price received from price discovery is lower than the expected price
-     *  - Reseller did not approve protocol to transfer exchange token in escrow
+     * - Offer price is in native token and caller does not approve WETH
+     * - Offer price is in some ERC20 token and caller also sends native currency
+     * - Calling transferFrom on token fails for some reason (e.g. protocol is not approved to transfer)
+     * - Received ERC20 token amount differs from the expected value
+     * - Transfer of voucher to the buyer fails for some reason (e.g. buyer is contract that doesn't accept voucher)
+     * - Current owner is not price discovery contract
+     * - Last voucher owner not found
+     * - Call to price discovery contract fails
+     * - Exchange doesn't exist after the call to price discovery contract
+     * - Exchange is not in the committed state
+     * - Price received from price discovery is lower than the expected price
+     * - Reseller did not approve protocol to transfer exchange token in escrow
+     * - New voucher owner is not buyer wallet
      *
-     * @param _exchangeId - the id of the exchange to commit to
-     * @param _exchangeToken - the address of the ERC20 token used for the exchange (zero address for native)
+     * @param _tokenId - the id of the token
+     * @param _offer - the fully populated BosonTypes.Offer struct
      * @param _priceDiscovery - the fully populated BosonTypes.PriceDiscovery struct
-     * @param _initialSellerId - the id of the original seller
      * @return actualPrice - the actual price of the order
      */
     function fulfilBidOrder(
-        uint256 _exchangeId,
-        address _exchangeToken,
+        uint256 _tokenId,
+        Offer storage _offer,
         PriceDiscovery calldata _priceDiscovery,
-        uint256 _initialSellerId
+        address owner,
+        ProtocolLib.ProtocolLookups storage _lookups,
+        IBosonVoucher bosonVoucher
     ) internal returns (uint256 actualPrice) {
-        IBosonVoucher bosonVoucher = IBosonVoucher(protocolLookups().cloneAddress[_initialSellerId]);
+        address priceDiscoveryContract = _lookups.priceDiscoveryContractByVoucher[_tokenId];
+
+        // If caller is not the owner, check if the owner exists and is voucher price discovery contract
+        require(owner != address(0) && owner == priceDiscoveryContract, OWNER_MUST_BE_PRICE_DISCOVERY_CONTRACT);
+
+        address exchangeToken = _offer.exchangeToken;
+
+        // Native token is not safe, as its value can be intercepted in the receive function.
+        if (exchangeToken == address(0)) exchangeToken = address(weth);
```

Comment:
If the offer is in the native token, we must also make sure that we unwrap ethers at the end, otherwise, there will be a mismatch in the escrow. (too much weth, not enough eth, and it would be impossible to finalize all offers in native tokens).